webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/facts/nodes/NodeFactRepository.scala

```diff
@@ -631,54 +631,60 @@ class CoreNodeFactRepository(
-                        s <- storage.save(updated)
-                        // then, we get the actual thing that was saved from the save event
-                        up = s match {
-                               case StorageChangeEventSave.Created(node, attrs)             =>
-                                 node
-                               case StorageChangeEventSave.Updated(oldNode, newNode, attrs) =>
-                                 newNode
-                               case StorageChangeEventSave.Noop(nodeId, attrs)              =>
-                                 updated
-                             }
-                        // then, we save that as the core node fact reference
-                        e <- nodeFact.rudderSettings.status match {
-                               case RemovedInventory  => // this case is ignored, we don't delete node based on status value
-                                 StorageChangeEventSave.Noop(up.id, attrs).succeed
-                               case PendingInventory  =>
-                                 saveOn(pendingNodes, up.toCore)
-                               case AcceptedInventory =>
-                                 saveOn(acceptedNodes, up.toCore)
-                             }
-                        // finally, merge both events and return the final change event and run call backs on it
-                        es = s.updateWith(e).toChangeEvent(up.id, up.rudderSettings.status, cc)
-                        _ <- NodeLoggerPure.Details.Write.trace(s"Saved node event: ${es.event.debugDiff}")
-                        _ <- runCallbacks(es)
-                      } yield es
-                    }
+        _            <- lock.withLock
+        core         <- get(node.fold(_.id, _._1))(cc.toQuery)
+        pair         <- (node, core) match {
+                          // first case: we are saving a new node fact, tenant can be set
+                          // we keep attrs from request
+                          case (Left(x), None)          => (x, attrs).succeed
+                          // second case: updating a node fact, we need to keep existing tenant
+                          // we also keep status
+                          // we keep attrs from request
+                          case (Left(x), Some(y))       =>
+                            (
+                              x
+                                .modify(_.rudderSettings.security)
+                                .setTo(y.rudderSettings.security)
+                                .modify(_.rudderSettings.status)
+                                .setTo(y.rudderSettings.status),
+                              attrs
+                            ).succeed
+                          // second case: we are updating tenants
+                          case (Right((_, t)), Some(x)) =>
+                            (NodeFact.fromMinimal(x.modify(_.rudderSettings.security).setTo(t)), SelectFacts.none).succeed
```

Comment:
we don't keep attributes from request ?

Reply:
in that case, we have a "right(tenant)`, so if the request is something else that "non", it's an error. I'm adding a comment